File: src/main/java/org/intellij/sdk/language/jass/JassFindUsagesProvider.java
package org.intellij.sdk.language.jass;

import com.intellij.lang.cacheBuilder.DefaultWordsScanner;
import com.intellij.lang.cacheBuilder.WordsScanner;
import com.intellij.lang.findUsages.FindUsagesProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiNamedElement;
import com.intellij.psi.tree.TokenSet;
import org.intellij.sdk.language.jass.psi.JassProperty;
import org.intellij.sdk.language.jass.psi.JassTokenSets;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

final class JassFindUsagesProvider implements FindUsagesProvider {

    @Override
    public @NotNull WordsScanner getWordsScanner() {
        return new DefaultWordsScanner(new JassLexerAdapter(),
                JassTokenSets.IDENTIFIERS,
                JassTokenSets.COMMENTS,
                TokenSet.EMPTY);
    }

    @Override
    public boolean canFindUsagesFor(@NotNull PsiElement psiElement) {
        return psiElement instanceof PsiNamedElement;
    }

    @Nullable
    @Override
    public String getHelpId(@NotNull PsiElement psiElement) {
        return null;
    }

    @NotNull
    @Override
    public String getType(@NotNull PsiElement element) {
        if (element instanceof JassProperty) {
            return "jass property";
        }
        return "";
    }

    @NotNull
    @Override
    public String getDescriptiveName(@NotNull PsiElement element) {
        if (element instanceof JassProperty) {
            return ((JassProperty) element).getKey();
        }
        return "";
    }

    @NotNull
    @Override
    public String getNodeText(@NotNull PsiElement element, boolean useFullName) {
        if (element instanceof JassProperty) {
            return ((JassProperty) element).getKey() +
                    JassAnnotator.SIMPLE_SEPARATOR_STR +
                    ((JassProperty) element).getValue();
        }
        return "";
    }

}

File: src/main/kotlin/raft/war/language/jass/JassFindUsagesProvider.kt
package raft.war.language.jass

import com.intellij.lang.cacheBuilder.DefaultWordsScanner
import com.intellij.lang.cacheBuilder.WordsScanner
import com.intellij.lang.findUsages.FindUsagesProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.tree.TokenSet
import raft.war.language.jass.psi.JassTypes.*

// https://plugins.jetbrains.com/docs/intellij/find-usages.html

internal class JassFindUsagesProvider : FindUsagesProvider {
    override fun getWordsScanner(): WordsScanner = DefaultWordsScanner(
        /* lexer = */ JassFlexAdapter(),
        /* identifierTokenSet = */ TokenSet.create(FUN_NAME),
        /* commentTokenSet = */ TokenSet.EMPTY,
        /* literalTokenSet = */TokenSet.create(FUN_NAME),
        /* skipCodeContextTokenSet = */ TokenSet.EMPTY,
        /* processAsWordTokenSet = */ TokenSet.create(FUN_NAME)
    )

    override fun canFindUsagesFor(psiElement: PsiElement): Boolean = psiElement is PsiNamedElement

    override fun getHelpId(psiElement: PsiElement): String? = null

    override fun getType(element: PsiElement): String {
        return ""
    }

    override fun getDescriptiveName(element: PsiElement): String {

        return ""
    }

    override fun getNodeText(element: PsiElement, useFullName: Boolean): String {
        return ""
    }
}